File: DFS和BFS/bfs 二进制矩阵中的最短路径.go
package DFS和BFS

type xyz struct {
	row        int
	col        int
	curPathLen int
}

//dirs描述了八个方向
var dirs = []struct{ i, j int }{{-1, 0}, {1, 0}, {0, -1}, {0, 1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}}

//从左上角开始，你可以走八个方向，能走的节点必须值为0，给出从左上角走到右下角的最短路径长度
func shortestPathBinaryMatrix(grid [][]int) int {
	n := len(grid)
	if n == 1 && grid[0][0] == 0 {
		return 1
	}
	var queue []*xyz
	if grid[0][0] == 0 {
		queue = append(queue, &xyz{0, 0, 1})
		grid[0][0] = 1 //表示0 0 位置已经走过了
	}
	for len(queue) != 0 {
		x := len(queue)
		for i := 0; i < x; i++ {
			cur := queue[0]
			queue = queue[1:]
			for j := 0; j < len(dirs); j++ { //八个方向各尝试走一次
				nextRow := cur.row + dirs[j].i
				nextCol := cur.col + dirs[j].j
				if nextRow >= 0 && nextRow < n && nextCol >= 0 && nextCol < n && grid[nextRow][nextCol] == 0 {
					queue = append(queue, &xyz{nextRow, nextCol, cur.curPathLen + 1})
					grid[nextRow][nextCol] = 1
					if nextRow == n-1 && nextCol == n-1 {
						return cur.curPathLen + 1
					}
				}
			}
		}
	}
	return -1
}

File: DFS和BFS/bfs二进制矩阵中的最短路径.java
package DFS和BFS;

import java.util.ArrayDeque;

public class bfs二进制矩阵中的最短路径 {
    public int shortestPathBinaryMatrix(int[][] grid) {
        int n = grid.length;
        if (grid[0][0] == 1 || grid[n - 1][n - 1] == 1) return -1;
        var queue = new ArrayDeque<int[]>();
        queue.add(new int[]{0, 0});
        grid[0][0] = 1;
        int step = 1;
        int[][] dirs = new int[][]{{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}};
        while (!queue.isEmpty()) {
            //表示第step步能达到的点有x个
            int x = queue.size();
            for (int i = 0; i < x; i++) {
                var cur = queue.poll();
                assert cur != null;
                int row = cur[0], col = cur[1];
                if (row == n - 1 && col == n - 1) return step;
                for (int k = 0; k < 8; k++) {
                    int nextRow = row + dirs[k][0], nextCol = col + dirs[k][1];
                    if (nextRow >= 0 && nextRow < n && nextCol >= 0 && nextCol < n && grid[nextRow][nextCol] == 0) {
                        queue.add(new int[]{nextRow, nextCol});
                        grid[nextRow][nextCol] = 1;
                    }
                }
            }
            step++;
        }
        return -1;
    }
}


class Day2023_5_26 {
    public int shortestPathBinaryMatrix(int[][] grid) {
        int n = grid.length;
        if (grid[0][0] == 1 || grid[n - 1][n - 1] == 1) return -1;
        int[][] DIRS = new int[][]{{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}};
        ArrayDeque<Integer> queue = new ArrayDeque<Integer>();
        queue.add(0);
        grid[0][0] = 1;
        int step = 0;
        while (!queue.isEmpty()) {
            step++;
            int x = queue.size();
            for (int i = 0; i < x; i++) {
                int cur = queue.poll();
                if (cur == n * n - 1) return step;
                int curX = cur / n, curY = cur % n;
                for (int[] D : DIRS) {
                    int nxtX = curX + D[0], nxtY = curY + D[1];
                    if (nxtX < 0 || nxtY < 0 || nxtX >= n || nxtY >= n || grid[nxtX][nxtY] != 0) continue;
                    grid[nxtX][nxtY] = 1;
                    queue.add(nxtX * n + nxtY);
                }
            }
        }
        return -1;
    }
}